kernel/src/main.rs

```diff
@@ -18,12 +19,22 @@ const PERFORM_ELF: &[u8] = include_bytes!(env!("CARGO_BIN_FILE_USER_perform"));
 
 #[no_mangle]
 #[inline(never)]
-extern "C" fn kmain() -> ! {
+extern "C" fn kmain(rb_in: RefCnt<RingBuffer>, rb_out: RefCnt<RingBuffer>) -> ! {
     log::info!("kmain");
     const ITERS: usize = 500;
     const N: usize = 1000;
     let mut cpu = CPU.borrow_mut();
 
+    let mut client = Client::new(rb_in, rb_out);
+    loop {
+        let msg = client.read_one_message();
+        let cont = client.process_incoming_message(msg, &mut cpu);
+        client.send_all();
+        if !cont {
+            break;
+        }
+    }
```

Comment:
This seems like the other half of the client's logic, which is not part of the definition of `Client` for some reason. Why does the message have to leave the client only to be sent back into it?